Add endpoint to move all replicas from a disk to another disk of the same broker

ilievladiulian opened the pull request:
This PR introduces a new endpoint that moves replicas from a specified disk to another disk of the same broker. This will allow the removal of the disk without the loss of data.

### What is new

The new endpoint is available as follows:
`POST /remove_disks?brokerid_and_logdirs=[brokerid1-logdir1,brokerid2-logdir2...]`

### What is missing

After all replicas are moved from the specified disk, the disk may still be used by CC during rebalances and Kafka can still use it when creating topics.

amuraru commented:
@ilievladiulian Please also add the new endpoint to https://github.com/adobe/cruise-control/tree/master/cruise-control/src/yaml

ctrlaltluc commented on cruise-control/src/yaml/endpoints/removeDisks.yaml:
```diff
@@ -0,0 +1,141 @@
+          type: string
+        example: "Balance disk utilization across all brokers in the cluster."
+      - name: brokerid_and_logdirs
+        in: query
```
A disk might have a lot of log dirs. If we want to remove the disk, and the disk has 1000 replicas (which is not much), each log dir having 50 chars, will we hit any URL limits having >50k chars (excluding protocol, host and other query params) in the URL string?

ctrlaltluc replied:
@ecojan thanks for clarifying! Thought we need to pass the path down to partition dir e.g. `/kafka-logs1/kafka/aep_pipeline_test_topic_1-0`, not disk path e.g. `/kafka-logs1`. I think there is no danger of URL length overflow then.

ilievladiulian replied:
As @ecojan mentioned, the parameter is not exactly a logdir, but a disk path to the logdirs. The endpoint moves all replicas from that path to another path configured on the same broker.

ctrlaltluc commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/DiskRemovalGoal.java:
```diff
@@ -0,0 +1,140 @@
+        }
+    }
+
+    @Override
+    public boolean optimize(ClusterModel clusterModel, Set<Goal> optimizedGoals, OptimizationOptions optimizationOptions) {
+        sanityCheckOptimizationOptions(optimizationOptions);
+
+        for (Map.Entry<Integer, Set<String>> brokerIdLogDirs : _brokerIdAndLogdirs.entrySet()) {
+            Integer brokerId = brokerIdLogDirs.getKey();
+            Set<String> logDirs = brokerIdLogDirs.getValue();
+
+            Broker currentBroker = clusterModel.broker(brokerId);
+            List<Disk> remainingDisks = new ArrayList<>();
+            currentBroker.disks().stream().filter(disk -> !logDirs.contains(disk.logDir())).forEach(remainingDisks::add);
+
+            int step = 0;
+            int size = remainingDisks.size();
+            while (!logDirs.isEmpty()) {
+                String logDirToRemove = (String) logDirs.toArray()[0];
+
+                Set<Replica> replicasToMove = currentBroker.disk(logDirToRemove).replicas();
+                while (!replicasToMove.isEmpty()) {
+                    Replica replica = (Replica) replicasToMove.toArray()[0];
+                    clusterModel.relocateReplica(replica.topicPartition(), brokerId, remainingDisks.get(step % size).logDir());
+                }
+
+                logDirs.remove(logDirToRemove);
+                step++;
+            }
+        }
+
+        return true;
+    }
```
```suggestion
    @Override
    public boolean optimize(ClusterModel clusterModel, Set<Goal> optimizedGoals, OptimizationOptions optimizationOptions) {
        sanityCheckOptimizationOptions(optimizationOptions);

        for (Map.Entry<Integer, Set<String>> brokerIdLogDirs : _brokerIdAndLogdirs.entrySet()) {
            relocateReplicas(clusterModel, brokerIdLogDirs);
        }

        return true;
    }

    private void relocateReplicas(ClusterModel clusterModel, Entry<Integer, Set<String>> brokerIdLogDirs) {
        Integer brokerId = brokerIdLogDirs.getKey();
        Set<String> logDirsToRemove = brokerIdLogDirs.getValue();

        Broker currentBroker = clusterModel.broker(brokerId);
        List<Disk> remainingDisks = currentBroker.disks().stream()
            .filter(currentDisk -> !logDirsToRemove.contains(currentDisk.logDir()))
            .collect(Collectors.toList());

        int logDirToRemoveCount = 0;
        int remainingDisksNumber = remainingDisks.size();
        for (String logDirToRemove : logDirsToRemove) {
            for (Replica replicaToMove : currentBroker.disk(logDirToRemove).replicas()) {
                clusterModel.relocateReplica(replicaToMove.topicPartition(), brokerId, remainingDisks.get(logDirToRemoveCount % remainingDisksNumber).logDir());
            }
            logDirToRemoveCount++;
        }
    }
```

ctrlaltluc replied:
Would the above suggestion work? What is different:
- the use of `while (!logDirs.isEmpty())` and `(String) logDirs.toArray()[0]` to iterate through a set looked very odd (same with the nested `replicasToMove` loop)
- some variable names could be more expressive (arguable, subjective), to make it easier to be read (e.g. `logDirs`, `step` or `size`)
- we reduce the complexity (a little) with all the nested loops, by extracting a method

It is optional, you decide.

ilievladiulian replied:
Refactor the code to improve readability. As a note, there are 2 nested loops there. The outside loop is refactored to use a for each iterator, however the inside loop calls a function which changes its iterator. This is why it is treated as a loop over a queue, so that it can change during the loop.

amuraru commented on cruise-control/src/yaml/endpoints/removeDisks.yaml:
```diff
@@ -0,0 +1,141 @@
+          default: true
+      - name: goals
+        in: query
+        description: List of goals used to generate proposal, the default goals will be used if this parameter is not specified.
```
how are the goals in thie endpoint being used? In my mind, moving data intra-broker wouldn't need inter-broker goals right?

ilievladiulian replied:
These parameters were required by the open api spec test because the RemoveDisksParameters class extended GoalBasedOptimizationParameters, but they were not used in the actual endpoint. Found a way to remove them and be ignored by the tests.

amuraru commented on cruise-control/src/yaml/endpoints/removeDisks.yaml:
```diff
@@ -0,0 +1,141 @@
+          additionalProperties:
+            type: array
+            items:
+              type: string
```
example?

ilievladiulian replied:
Added an example in the latest commit (ex: 1-/tmp/kafka-logs-1)

amuraru commented on cruise-control/src/yaml/endpoints/removeDisks.yaml:
```diff
@@ -0,0 +1,141 @@
+          type: string
+        example: "Balance disk utilization across all brokers in the cluster."
+      - name: brokerid_and_logdirs
+        in: query
```
is this required?

ilievladiulian replied:
Yes, the parameter is required. Updated spec to reflect that.

amuraru commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/RemoveDisksRunnable.java:
```diff
@@ -0,0 +1,165 @@
+    protected final Map<Integer, Set<String>> _brokerIdAndLogdirs;
+
+    /**
+     * Constructor to be used for creating a runnable for self-healing.
```
stale java doc?

ilievladiulian replied:
Yes, it was a java doc for a constructor which is not used, just copy-paste leftover from another runnable.

amuraru commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/RemoveDisksRunnable.java:
```diff
@@ -0,0 +1,165 @@
+                               String anomalyId,
+                               Supplier<String> reasonSupplier,
+                               boolean stopOngoingExecution) {
+        super(kafkaCruiseControl, new OperationFuture("Broker Removal for Self-Healing"), selfHealingGoals, allowCapacityEstimation,
```
`Broker Removal for Self-Healing` reads confusing here. Can you expand of that?

ilievladiulian replied:
The code is from a constructor which is not used (copy-paste leftover). Removed it in latest commits.

ctrlaltluc commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/RemoveDisksRunnable.java:
```diff
@@ -0,0 +1,165 @@
+            Integer brokerId = entry.getKey();
+            Set<String> logDirs = entry.getValue();
+            Broker broker = clusterModel.broker(brokerId);
+            if (broker.disks().size() < logDirs.size()) {
```
Should we validate against erroneous log dir names too, not only size?

ilievladiulian replied:
Added validation against configured log dir names as well.

ctrlaltluc replied:
Nice, would also differentiate the error messages when the broker dir names are an issue versus when the broker dir number is an issue. Right now the error messages are the same.

ilievladiulian replied:
Changed log message to differentiate between issues.

amuraru commented:
@ilievladiulian can you please also add unit and integration tests for this new feature?

ctrlaltluc commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/RemoveDisksRunnable.java:
```diff
@@ -0,0 +1,165 @@
+                    remainingCapacity += disk.capacity();
+                }
+            }
+            if (removedCapacity > remainingCapacity) {
```
Should we also add a margin of error here? Edge case, if the capacities are equal, it would still be an issue, as it would result in full disks. Maybe a 5-10%? cc @amuraru

ilievladiulian replied:
Added an error margin check there, which is configurable.

ctrlaltluc reviewed (commented):
Nice! 
Left some questions, most importantly about URL length limitations and disk capacity validations.

ilievladiulian commented:
Added a unit test for the newly created goal, DiskRemovalGoalTest which checks that replicas from a logdir are moved to another logdir on the same broker.

Also added a configurable parameter for the margin of error when checking that the remaining disk size is enough to move the requested disks to remove. The option is configurable in the cruisecontrol properties file under the `remove.disks.remaining.size.error.margin` name. The default value is set to `0.1` which translates into `(removedSize / remainingSize) > (1 - 0.1)`

ctrlaltluc commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/DiskRemovalGoal.java:
```diff
@@ -0,0 +1,143 @@
+        for (String logDirToRemove : logDirsToRemove) {
+            Set<Replica> replicasToMove = currentBroker.disk(logDirToRemove).replicas();
+            while (!replicasToMove.isEmpty()) {
+                Replica replica = (Replica) replicasToMove.toArray()[0];
```
```suggestion
                Replica replica = replicasToMove.iterator().next();
```

ilievladiulian replied:
Refactored.

ctrlaltluc commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/RemoveDisksRunnable.java:
```diff
@@ -0,0 +1,154 @@
+                }
+            }
+            double errorMargin = (double) _kafkaCruiseControl.config().mergedConfigValues().get(REMOVE_DISKS_REMAINING_SIZE_ERROR_MARGIN);
+            if (removedCapacity / remainingCapacity > (1 - errorMargin)) {
```
This is not enough as validation. We need to compare actual disk utilization of the remaining disks (if we were to evict the removed disks) with actual disk capacity of the remaining disks.
So the condition should be smth like:
```
if ( 1 - (remainingUsageAfterRemoval / remainingCapacity) > errorMargin)
```
where:
```
remainingUsageAfterRemoval = current usage for remaining disks + additional usage from removed disks
```

**This condition does not cover all edge cases either**, as we might have (to the extreme) a removed disk with a single very large partition, that might not fit on one remaining disk, although the condition above is respected. So the validation should be also done when moving. 

I would check other goal validations too, for inspiration, to see how it was handled.

ilievladiulian replied:
Changed the validation to use disk utilization from removed replicas and current disk to cover this case.

ctrlaltluc reviewed (commented):
Left one more main comment on disk usage validation. After that is solved, I will approve. Thanks for your patience!

ctrlaltluc commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/DiskRemovalGoal.java:
```diff
@@ -0,0 +1,143 @@
+                clusterModel.relocateReplica(
+                        replica.topicPartition(),
+                        brokerId,
+                        remainingDisks.get(removedLogDirsCount % remainingDisksNumber).logDir()
```
There is one more possible issue here, I missed it in my first review iterations. Say that before removing disks we have:
```
/disk1 10% free
/disk2 10% free
/disk3 100% free
```
and the space free is enough to accommodate the replica movements.

With the current shuffle logic (modulo remaining disks), we will have an issue moving the replicas to disk 1, as it is already "almost full". However we can move all the replicas to disk 3.

Maybe we do not have this use case in Adobe, but if we plan to contribute upwards, we need to cover this IMO.

ilievladiulian replied:
Added also a validation for disk usage when relocating the replica. The loop now checks each "remaining disk" and moves the replica to the first one that has enough space.

ctrlaltluc commented on cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/DiskRemovalGoalTest.java:
```diff
@@ -0,0 +1,148 @@
+    }
+
+    @Test
+    public void testMoveLargeDiskToEmptyDisk() {
```
I think the first 3 tests can become 1 test, when there is enough space to move them.

ilievladiulian replied:
Merged tests into a single one.

ctrlaltluc commented on cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/DiskRemovalGoalTest.java:
```diff
@@ -0,0 +1,148 @@
+
+    @Test
+    public void testReplicasStayIfDiskAlmostFull() {
+        ClusterModel clusterModel = createClusterModel();
```
I think the last 2 tests can become 1 test, when there isn't enough space to move them.

ilievladiulian replied:
Merged tests into a single one.

ctrlaltluc commented on cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/DiskRemovalGoalTest.java:
```diff
@@ -0,0 +1,148 @@
+import static com.linkedin.kafka.cruisecontrol.config.constants.AnalyzerConfig.DEFAULT_REMOVE_DISKS_REMAINING_SIZE_ERROR_MARGIN;
+import static org.junit.Assert.assertEquals;
+
+public class DiskRemovalGoalTest {
```
I feel much safer with tests added! 
Some suggestions:
- add assertions for all log dirs, not only the one to remove
- add another test when there is more than 1 extra disk to move replicas to (to highlight the moving behavior that they will be filled up in order, not uniformly - as a side-note, if it was my implementation, I would fill them round-robin, if they have space, otherwise a new intra-broker rebalance needs to be triggered after removing the disk). Your call, it will probably come up when making the PR in upstream.

ilievladiulian replied:
Added assertions for all log dirs involved in the test.
Changed the replica distribution when relocating from filling up in order to round-robin. Also added tests for this.

ctrlaltluc commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/DiskRemovalGoal.java:
```diff
@@ -0,0 +1,169 @@
+            for (int i = 0; i < replicasToMove.size(); i++) {
+                Replica replica = replicasToMove.iterator().next();
+                relocateReplicaIfPossible(clusterModel, brokerId, remainingDisks, replica);
+                LOG.info(String.format("Could not move replica %s to any of the remaining disks.", replica));
```
SLF4J loggers support placeholders for strings with the following syntax:
```suggestion
                LOG.info("Could not move replica {} to any of the remaining disks.", replica);
```

ilievladiulian replied:
That's cool! Didn't know about it. Removed call to string format.

amuraru commented on cruise-control/src/yaml/endpoints/removeDisks.yaml:
```diff
@@ -0,0 +1,69 @@
+            items:
+              type: string
+        required: true
+        example: 101-/tmp/kafka-logs-1,101-/tmp/kafka-logs-2
```
nit:
```suggestion
        example: 101:/tmp/kafka-logs-1,101:/tmp/kafka-logs-2
```
can we use `:` as a separator between brokerid and logdir? reads more cleaner

ilievladiulian replied:
I believe it is possible to do it so. However, I used the same format as for the broker id and logdir param from the demote broker endpoint, for consistency purposes.

amuraru reviewed (approved):
Solid work @ilievladiulian

amuraru commented on cruise-control/src/yaml/endpoints/removeDisks.yaml:
```diff
@@ -0,0 +1,69 @@
+RemoveDisksEndpoint:
+  post:
+    operationId: removeDisks
+    summary: Move all replicas from the specified disk to another disk of the same broker.
```
```suggestion
    summary: Move all replicas from the specified disk to other disks of the same broker.
```

It's not only one disk where we move the replicas to, right?

amuraru commented:
@ilievladiulian one thing that's not very clear to me is whether the removed/decommissioned disks are kept as such (disabled) for a well defined period.
Marking them as decomissioned is important to ensure a subsequent rebalanace operation (via automated self-healing or manual cluster rebalance) is not bringing back data to removed disks.

Can you confirm whether this functionality is encoded in this PR?

ecojan reviewed (approved):
Looks good! Great job @ilievladiulian

ilievladiulian commented:
@amuraru the disks are not marked as disabled after this operation. I identified no trivial way to do this and after discussing internally we convened to use this functionality as it is right now with the following workflow:

1. call remove disks endpoint to move the replicas from the specified disks
2. after the operation finishes, put CC in maintenance mode so no rebalance occurs
3. manually delete the disk

As a note, even if we are able to mark the disks as disabled and no rebalance will run, the disks can still be used by Kafka if a new topic is created.
